### ISSUE TITLE
smote did not work for small number of cases

### ISSUE THREAD

**Issue by yfan18** [s-kganz/scutr#1]
Here is the number of cases in each class (total 22 classes, 460 samples, average is 21)
  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20
  3   4   2  44  27  10   7  15  15  10   5   8 115   8   3  46  38   6   4  58
 21  22
  3  29
I ran scut usng default.
command:  scutted <- SCUT(zz, "class")
I get the following error:
Error in get.knnx(data, query, k, algorithm) : ANN: ERROR------->
Calls: SCUT ... <Anonymous> -> SMOTE -> knearest -> <Anonymous> -> get.knnx
In addition: Warning message:
In get.knnx(data, query, k, algorithm) : k should be less than sample size!

It seems for SMOTE to work,  K cannot be greater than or equal to the sample size. (in my case it's 2).  But to generate 20 samples, k have to >2.   is there a way to get around this? Thanks!

**Comment by s-kganz** [comment]
Hello,

Thanks for pointing this out. SMOTE works by picking k neighbors for a particular point, then it interpolates new points between them. I didn't expose the k parameter in the old version. `oversample_smote` now picks a value for k that will work for small sample sizes and throws an error if the user gives one that is too large. Here is a reprex to verify that it works:
```
library(scutr)

n <- as.numeric(
  strsplit("3 4 2 44 27 10 7 15 15 10 5 8 115 8 3 46 38 6 4 58 3 29", " ")[[1]]
)

class <- c()
for (i in 1:length(n)) {
  class <- c(class, rep(i, n[i]))
}

# Make some data
set.seed(1234)
ncol <- 2
X <- matrix(runif(2 * length(class)), ncol=2)
df <- as.data.frame(cbind(class, X))

scutted1 <- SCUT(df, "class")

small <- scutted1[scutted1$class == 3, ]
plot(small$V2, small$V3)
```
I kind of question the validity of making synthetic cases for a class with two observations. At that point I would just go look for more data :). As you can see, SMOTE just adds new points on a line between the two.

Let me know if you have any issues. I can upload this to CRAN later today if all is well.

**Comment by yfan18** [comment]
Sounds good! Thanks for your quick response！Sent from my iPhoneOn Nov 10, 2023, at 11:48 AM, Keenan Ganz ***@***.***> wrote:﻿
Hello,
Thanks for pointing this out. SMOTE works by picking k neighbors for a particular point, then it interpolates new points between them. I didn't expose the k parameter in the old version. oversample_smote now picks a value for k that will work for small sample sizes and throws an error if the user gives one that is too large. Here is a reprex to verify that it works:
library(scutr)

n <- as.numeric(
  strsplit("3 4 2 44 27 10 7 15 15 10 5 8 115 8 3 46 38 6 4 58 3 29", " ")[[1]]
)

class <- c()
for (i in 1:length(n)) {
  class <- c(class, rep(i, n[i]))
}

# Make some data
set.seed(1234)
ncol <- 2
X <- matrix(runif(2 * length(class)), ncol=2)
df <- as.data.frame(cbind(class, X))

scutted1 <- SCUT(df, "class")

small <- scutted1[scutted1$class == 3, ]
plot(small$V2, small$V3)

I kind of question the validity of making synthetic cases for a class with two observations. At that point I would just go look for more data :). As you can see, SMOTE just adds new points on a line between the two.
Let me know if you have any issues. I can upload this to CRAN later today if all is well.

**Comment by s-kganz** [comment]
Fixed by #2
